cataclysmbnteam/Cataclysm-BN
# Weapon Categories for gunmods.

#### Summary

SUMMARY: Infrastructure "Restructure gunmod eligibility functions."

#### Purpose of change

- #1709  is merged so we can start to restructure gun_mods to use weapon categories as targets.

This should cut down on the weird hacks we have with crossbow needing hardcode to translate, and hand crossbows being the only type to require 2 mod_targets to be satisfied.

#### Describe the solution

First tears out gun_type_type because that was a weird thing that stored both skill ids and itype_ids. Made a new usable_category field that will contain sets of weapon categories which should allow you to specify gunmods to require more than one field to be satisfied, while still supporting other single category filters.

#### Describe alternatives you've considered

Merge weapon_category_id into usable field.
- Sounds horrifying.

#### Testing

- Add PISTOL category to `mod_target_category` of `belt_clip`. Spawn Mossberg Brownie, RM103A automagnum, v29 laser pistol, A7 laser rifle
- [x] Check that belt clip can be attached to Mossberg Brownie, RM103A automagnum and v29 laser pistol only.
  - Confirms that categories and exceptions do not conflict.
- [x] Check that Focusing Lens can be attached to v29 laser pistol and A7 laser rifle only
  - Confirm that multiple categories are parsed appropriately.
- [x] Check that electrolaser conversion can only be attached to v29 laser pistol.
  - Further confirm that multiple categories don't have weird behaviour.

#### Additional context
